### DISCUSSION TITLE
Enjoying BrowseRouter as Browser Tamer replacement

### DISCUSSION THREAD

**Discussion by freiheitner** [nref/BrowseRouter#14] [General]
I am glad to have found BrowseRouter. I'm enjoying it as Browser Tamer replacement.  BT offers a user interface to manage which domains open to which browser, but it's starting to suffer feature creep and occasionally trips antivirus alerts.  BrowseRouter is simple and does one thing well.

**Comment by nref** [comment]
Thanks for the positive feedback!

What features of Browser Tamer have turned you away? I ask because I have thought about adding a UI since only a fraction of users may be comfortable editing an INI file setting up BrowseRouter the command line. I would keep the CLI around of course.

**Reply by freiheitner** [reply]
I see aloneguid has replied on this thread and kind of covers my concerns -- I was not looking for a CPU, RAM and disk use monitor in an app which primarily runs in the background routing my browser requests.  But that was okay until Windows Defender (or whatever MS calls it this month) flat out deleted Browser Tamer from my system as a virus during the night a few weeks ago.  It had previously warned me about it, but I told it that BT was alright, and then it went and deleted it anyway.  I was looking for something--anything--to serve the same simple purpose of launching different browsers and browser profiles by the URL without raising any fuss with anti-malware tools.

I do love BT's interface for assigning URLs to different browsers and that BT auto-detects different browser profiles.  I was able to get those working via manual editing in BrowseRouter's config file.

**Comment by nref** [comment]
Interesting development: @aloneguid [open-sourced](https://github.com/aloneguid/bt) Browser Tamer yesterday.

**Reply by aloneguid** [reply]
Yes for the reasons above ) antivirus doesn't like it sometimes due to automatically discovering browsers. I've got nothing to hide and decided to open source it. Also removed features already in 3.1 @freiheitner - I realised myself just after adding them that better to do one thing well rather than bolt on some unrelated rubbish (c) UNIX.

**Reply by nref** [reply]
I read your blog post about your eponymous PUA. Seems pernicious to me. Open-sourcing may be the best move if you don't want to monetize it.

**Reply by aloneguid** [reply]
No, I don't think of monetising in any way. Closed source was due to not wanting to deal with overhead of pull requests and questions. I already do feel like spending more time on community rather that enjoying implementing features. So nothing is perfect.

**Reply by flawrist** [reply]
How to prevent BrowseRouter from polluting the disk with logs?

**Reply by nref** [reply]
> How to prevent BrowseRouter from polluting the disk with logs?

Not supported currently. Please open an issue and I will add support.

**Reply by nref** [reply]
> > How to prevent BrowseRouter from polluting the disk with logs?
> 
> Not supported currently. Please open an issue and I will add support.

This is now supported in the new `[log]` config section. See the README.

**Comment by clayissimo** [comment]
> I am glad to have found BrowseRouter. I'm enjoying it as Browser Tamer replacement. BT offers a user interface to manage which domains open to which browser, but it's starting to suffer feature creep and occasionally trips antivirus alerts. BrowseRouter is simple and does one thing well.

@freiheitner - just curious - do you see any important differences between BrowserRouter and Browser Tamer now? I just stumbled across Browser Tamer and both tools seem to do the same thing.

**Comment by nref** [comment]
Here's a difference: your privacy.

According to the author, BrowserTamer had 15,000 installs in 2024. I cannot tell you how many BrowseRouter has because *I do not track it*.  BrowseRouter has no analytics or tracking whatsoever, and it never will because I would not want to be tracked, and I treat my users the way I want to be treated.

If you'd like a browser middleware app that has a privacy commitment, a nice GUI, works on Windows, macOS, and Linux, and my devoted support, try my app [Linklever](https://linklever.net/). It's free to try. When not licensed, it will occasionally remind you to buy, but it has no limitations.

![image](https://github.com/user-attachments/assets/f9d10583-adb2-467f-8c4e-9d9797f385d7)

**Comment by clayissimo** [comment]
@nref - privacy! Good point! Thank you for your input.

From what I understand (haven't checked the sources) it only tracks once per installation. If this is true, personally, I do not see any major issue with this, since it does not permanently track/collect information.

I'd like to know whether there are any major differences functionality wise.

**Reply by aloneguid** [reply]
Actually, you don't know if you are tracked, because Linklever is not OSS. In fact, Linlever does phone back home on startup, so it is tracking you (another marketing lie?). Browser Tamer's source, build and release are fully open, and you can clearly see there is zero tracking. Browser Tamer also does not limit you to a number of rules and does not charge anything – it's a hobby project made to solve one problem, and the author actually uses it daily for this purpose. There is no aim to earn money on it. Also, the number of features is probably at least 100 times more. Browser Tamer also does not install another browser on your computer (it's not Electron) and is 100% native. It does not hide the fact it's Windows only, and proud of it – working well on one OS rather than working badly on all of them. You should check the facts yourself and not blindly believe anything written on the internet ;)

**Reply by clayissimo** [reply]
> Actually, you don't know if you are tracked, because Linklever is not OSS.

@aloneguid - spare your breath. You're off topic. You might have missed it: This discussion (and esp. my comment) is about BrowserRouter, not about LinkLever.

> You should check the facts yourself

... and not go off topic. Thank you.

**Reply by aloneguid** [reply]
It's a bot

**Reply by nref** [reply]
@clayissimo thank you for the backup. I suppose I did invite discussion of Linklever by bringing it up 11 months ago.

@aloneguid I can see how my statement on privacy could have agitated you. However, your tone is not appropriate here, and your statements are not factual. 

> another marketing lie

Where was the first marketing lie? 

> Linlever does phone back home on startup

It does not. Linklever checks for updates if allowed in Settings. If you would kindly share what network requests caused you concern, I will take action to rectify them.

> Browser Tamer also does not install another browser on your computer (it's not Electron) and is 100% native

This is all true of BrowseRouter and Linklever. No browser, not Electron. It's AOT compiled C#, so a "native" app in the sense that it's bytecode for your OS and architecture.

> Linklever is not OSS

Never said it was. Charging $29 lets me give it "my devoted support". I add features requested by customers nearly monthly. For example, just added Hyprland support on Linux and the Ask hotkey on all platforms. Sometimes I [blog](https://linklever.net/blog/) about them.

**Reply by aloneguid** [reply]
@nref All I did is uploaded the latest .exe of LinkLevel to VirusTotal, the sandbox showed it does phone home to check for updates. This is tracking. Saying "no it does not" is not factual. Simply checking web server logs gives you the approximate amount of installations. I really don't mind this, but what's the point of lying if this can be checked trivially? I wonder what else you are not disclosing.

"If allowed in settings" - 'cmon, there is no way to turn this off. It's enabled by default and opening the application first time even to turn it off constitutes a home call. Why are you disrespecting users like that and talking about privacy? Not everyone is stupid you know.

**Reply by nref** [reply]
I waited three weeks to reply so I could introspect first. You called me a liar, and you've suggested that I think my users are "stupid" and that I've disrespected them. These are ad hominem attacks on my character, which I take seriously because I follow Jesus, and I care deeply how I reflect on him. So rather than just redact or delete your inflammatory accusations, my hope is to address them honestly and gain your trust and that of those who happen to land here later.

Looking back up this chat, this all started when I claimed about BrowseRouter (not Linklever) that it "has no analytics or tracking whatsoever", which is true. I contrasted this to BrowserTamer does seem to count the number of installs. I then said "I treat my users the way I want to be treated" which I see now implied that you do not. I apologize for that insinuation. It was an overreach.

Back to Linklever. In this chat, I never claimed it "has no analytics or tracking whatsoever", merely that it "has a privacy commitment". However, the landing page claims "Completely local and offline. Zero tracking." The spirit of this marketing line is in keeping with daily operations of the app, which are indeed local and offline. However, it does merit some qualification since nothing that is "completely offline" can check for updates. But regarding updates, I'll infer that you did not run the app. You claimed "there is no way to turn this off". Had you run it, you would have seen the toggle in Settings.

<img width="302" height="91" alt="image" src="https://github.com/user-attachments/assets/acce72bb-d097-4bd9-b13a-260279a68daf" />

At this point I think it's important to define "phone home" and "tracking". I understand "phone home" to be a derogatory ET-flavored metaphor for software that sends data back to the developer. "Tracking" is the collection of data over time that includes some kind of stable identifier, like a person, device, or browser.

Regarding "phoning home". You don't have to take my word on what Linklever sends. Point a proxy at it (e.g. mitmproxy, Wireshark) and you'll see on the `GET` there's no telemetry or payload.

Regarding "tracking", I'd like to identify some assumptions you made.

- Checking for updates sends some kind of stable identifier ("This is tracking")
- That Linklever updates are served via a web server whose logs I can check ("checking web server logs")

Regarding stable identifiers, you were right, but this was inadvertent on my part and I am removing it. The update mechanism in Linklever is provided by [Velopack](https://github.com/velopack/velopack). To update, it sends a request like this:

  `GET /releases/win-x64/releases.win.json?localPackage=<version>&id=<stagedUserId>`

where `id` is a random GUID generated once per install and stored locally. I didn't know this query parameter existed until a few minutes ago when I did some research for this reply. 

Regarding web server logs, all Linklever downloads, including the version manifest file, are served from Cloudflare CDN and originate in R2 storage. Anyone can verify the site sits behind Cloudflare:
  
```
$ curl -sI https://linklever.net/
HTTP/2 200
...
server: cloudflare

$ dig linklever.net
...
linklever.net.          300     IN      A       104.21.79.153
linklever.net.          300     IN      A       172.67.146.115

$ whois 104.21.79.153
OrgName:        Cloudflare, Inc.
NetName:        CLOUDFLARENET
CIDR:           104.16.0.0/12

$ whois 172.67.146.115
OrgName:        Cloudflare, Inc.
NetName:        CLOUDFLARENET
CIDR:           172.64.0.0/13

$ curl -s https://rdap.arin.net/registry/ip/104.21.79.153
...
"fn", "text", "Cloudflare, Inc."
```

So there is no web server. Now, it is possible to set up Cloudflare to track requests, but I haven't. I can't prove that to you, so that's where I'll leave it.

Here's what I am doing in response to your concerns:

- I've updated the privacy policy to disclose auto-update and spell out what data Linklever collects. Short version: nothing. https://linklever.net/privacy
- I've more clearly disclosed auto-update on the landing page.
- As of `0.65.2`, I've removed the Velopack `id` query parameter, which did not know about and I have never used.

Beyond that, I'd like to ask **what else you would have me do**. I'll say up front what I can't do:

- Make Linklever open source.
- Concede that OSS is categorically more private or secure just because the source is public.
- Make auto update off by default. That has support tradeoffs I can't accept, and none of my paying customers are asking for it.
- Give you access to my git forge or Cloudflare to verify my claims. It wouldn't prove anything, since I could change everything after you look. After all, you're not stupid.

In short, [Hanlon's razor](https://en.wikipedia.org/wiki/Hanlon%27s_razor) applies here. You've attributed malice to me, but this whole thing was my fault for picking a fight about privacy.

Thanks for helping Linklever live up to its mission of being private browser middleware.
